### DISCUSSION TITLE
Rebuild segment OBR.28 from variable

### DISCUSSION THREAD

**Discussion by Frazou1** [nextgenhealthcare/connect#6072] [Q&A]
Hi, so my goal is to read the orignal message obr28 segment. On that segment i have multiple occurance. Im able to identified each one and put on a variable.  After that i do a copy of hl7 and remove de obr28 segment before recreate the new one with the new value. So on that script i want to generate and send new HL7 for each occurance in the obr28 segment. The problem i have, i not able tu put new value on the message.      See the code :   

// Accéder au segment OBR
// Fonction récursive pour une copie profonde d'un objet
function deepCopy(obj) {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }

    if (obj instanceof Array) {
        var copyArr = [];
        for (var i = 0; i < obj.length; i++) {
            copyArr[i] = deepCopy(obj[i]);
        }
        return copyArr;
    }

    if (obj instanceof Object) {
        var copyObj = {};
        for (var key in obj) {
            if (obj.hasOwnProperty(key)) {
                copyObj[key] = deepCopy(obj[key]);
            }
        }
        return copyObj;
    }
}

var newHL7Messages = [];

var obrSegment = msg['OBR'];

// Initialiser une variable pour stocker toutes les occurrences des médecins en CC
var obrData = [];

// Afficher la structure du message
logger.info('Message Structure: ' + msg.toString());

if (obrSegment) {
    logger.info('Entered the OBR condition');
    logger.info('OBR Segment Size: ' + obrSegment.length());
    logger.info('OBR Segment Content: ' + obrSegment.toString());

    // Ajout d'une liste pour stocker les informations OBR.28
    var obr28Data = [];

    for (var i = 0; i < obrSegment.length(); i++) {
        var subSegment = obrSegment[i];
        logger.info('Processing OBR sub-segment: ' + subSegment.name().toString());

        // Ajoutez ce log pour voir le contenu du sous-segment OBR
        for (var j = 0; j < subSegment.length(); j++) {
            logger.info("Sub-segment name: " + subSegment[j].name().toString());
            logger.info("Sub-segment name2: " + subSegment['OBR.28'].toString());

        }

         if (subSegment.name() == "OBR" && subSegment['OBR.28']) {
            logger.info('Found OBR.28 sub-segment');
            var obr28Array = [];
            var obr28Occurrences = subSegment['OBR.28'];
            for (var j = 0; j < obr28Occurrences.length(); j++) {
                // Modifiez les valeurs selon vos besoins
                //obr28Occurrences[j]['OBR.28.1'] = 'NouvelleValeur1';
                //obr28Occurrences[j]['OBR.28.2'] = 'NouvelleValeur2';
                // ... Ajoutez d'autres modifications nécessaires
			var newHL7Message = {};
			 var newHL7Message = deepCopy(msg);

			delete newHL7Message['OBR']['OBR.28'][0];
			 
                var obrInfo = {};
                obrInfo['CCID'] = obr28Occurrences[j]['OBR.28.1'].toString();
                obrInfo['CCLastName'] = obr28Occurrences[j]['OBR.28.2'].toString();
                obrInfo['CCFirstName'] = obr28Occurrences[j]['OBR.28.3'].toString();
                obrInfo['Fax'] = obr28Occurrences[j]['OBR.28.16']['OBR.28.16.1'].toString();
                obrInfo['ClinicName'] = obr28Occurrences[j]['OBR.28.16']['OBR.28.16.2'].toString();

        // Ajoutez les informations OBR à la liste
        obrData.push(obrInfo);

        // Affichez les informations pour cette occurrence
        logger.info('OBR Info ' + obrInfo['ClinicName']);
		    
		    var newObr28Occurrence = {};
		    newObr28Occurrence['OBR.28.1'] = obrInfo['CCID'];
		    newObr28Occurrence['OBR.28.2'] = obrInfo['CCLastName'];
		    newObr28Occurrence['OBR.28.3'] = obrInfo['CCFirstName'];
		    newObr28Occurrence['OBR.28.16'] = {
		        'OBR.28.16.1': obrInfo['Fax'],
		        'OBR.28.16.2': obrInfo['ClinicName']
		    };

	        obr28Array.push(newObr28Occurrence);
        		newHL7Message['OBR']['OBR.28'] = obr28Array;
                // Convertir le message HL7 modifié en une chaîne et afficher dans la console
                var newHL7String = newHL7Message.toString();
                logger.info('Nouveau message HL7: ' + newHL7String);

            }
        }
    }
    // Utilisez la liste des informations OBR.28 selon vos besoins
    logger.info('OBR.28 Data: ' + JSON.stringify(obr28Data));

}

**Comment by pacmano1** [comment]
Can you just provide a before and after message?  Reading someone's code is way harder.

**Comment by Frazou1** [comment]
Of course, so Original message segment obr28 look like this :  `<OBR.28><OBR.28.1>111322</OBR.28.1><OBR.28.2>JUTRAS</OBR.28.2><OBR.28.3>JEAN-FRANCOIS</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8197777777</OBR.28.16.1><OBR.28.16.2>STE-GERTRUDE</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28><OBR.28><OBR.28.1>111322</OBR.28.1><OBR.28.2>RICHARD</OBR.28.2><OBR.28.3>JEAN</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8193666666</OBR.28.16.1><OBR.28.16.2>CLINIQUES Interne</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28><OBR.28><OBR.28.1>101708</OBR.28.1><OBR.28.2>PRIEM</OBR.28.2><OBR.28.3>MARINE</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8195555555</OBR.28.16.1><OBR.28.16.2>SANTE M</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28>
`

The result i want is :MESSAGE 1 `<OBR.28><OBR.28.1>111322</OBR.28.1><OBR.28.2>JUTRAS</OBR.28.2><OBR.28.3>JEAN-FRANCOIS</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8197777777</OBR.28.16.1><OBR.28.16.2>STE-GERTRUDE</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28>` MESSAGE 2 `
<OBR.28><OBR.28.1>111322</OBR.28.1><OBR.28.2>RICHARD</OBR.28.2><OBR.28.3>JEAN</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8193666666</OBR.28.16.1><OBR.28.16.2>CLINIQUES Interne</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28>` MESSAGE 3 `
<OBR.28><OBR.28.1>101708</OBR.28.1><OBR.28.2>PRIEM</OBR.28.2><OBR.28.3>MARINE</OBR.28.3><OBR.28.4/><OBR.28.5/><OBR.28.6/><OBR.28.7/><OBR.28.8/><OBR.28.9/><OBR.28.10/><OBR.28.11/><OBR.28.12/><OBR.28.13/><OBR.28.14/><OBR.28.15/><OBR.28.16><OBR.28.16.1>8195555555</OBR.28.16.1><OBR.28.16.2>SANTE M</OBR.28.16.2><OBR.28.16.3>L</OBR.28.16.3></OBR.28.16></OBR.28>`

**Reply by pacmano1** [reply]
Sorry I mean a before and after HL7 message, not XML.  A full before and after message.

**Reply by Frazou1** [reply]
I can provide you my message before but not the after because im not able to generate it ...  ` MSH|^~\&|SRTEXT|Neuron^Radiologie^L|DME|Neuron^Patate^L|20240114161319||ORU^R01|011111|P|2.5.1|||AL||CA|8859/1
PID|1||0003885|AAAA11111111|TEST^TEST|^|19460101|F|||2020, DU TEST^^SAINT^QuÃ©bec^H0H 0H0||(819) 999-9999
PV1|1|O|||||199460^TEST2^TEST2^^^^^^^^^^^^^8198888888&Tomate&L|107133^TEST3^TEST3^^^^^^^^^^^^^8195555555&Patate&L
ORC|NW||20231219-169||CM|||||||107133^TEST3^TEST3||||FINAL
OBR|1||20231219-169|8049^MAMMO||20231219141400|20231219142300||||||||SE&Sein|107133^TEST3^TEST3||20231219-169|20231219-16901|||20240103115058||MG|F|||111322^TEST4^TEST4^^^^^^^^^^^^^8192222222&STE-GERTRUDE&L~111322^TEST5^TEST5^^^^^^^^^^^^^8192222222&CLINIQUES Interne&L~101708^TEST6^TEST6^^^^^^^^^^^^^8191111111&CLINIQUE DE SANTE M&L||||1126333&DIEZ MARTINEZ&PATRICIA||BAMA01&BAZIN-CHAMPAGNE&MAUDE|MATJOS01&MATTEAU&JOSÃ‰E-ANNE|20231219141400 
`

**Reply by pacmano1** [reply]
LOL, FAKE IT.  _I mean post the message as it should look if it did work._

**Reply by pacmano1** [reply]
Also please read up on highlighting your code, you are doing it wrong.

You just need to start the code with three back ticks and the letter "JS" and end it with three back ticks:

```JS
var x = 1
```

In the editor before I post the response that looks like:
![image](https://github.com/nextgenhealthcare/connect/assets/44065187/8fefd5ca-854f-4f7f-8754-f3015350dd46)

**Reply by pacmano1** [reply]
Your HL7 message:

```TEXT
MSH|^~\&|SRTEXT|Neuron^Radiologie^L|DME|Neuron^Patate^L|20240114161319||ORU^R01|011111|P|2.5.1|||AL||CA|8859/1
PID|1||0003885|AAAA11111111|TEST^TEST|^|19460101|F|||2020, DU TEST^^SAINT^QuÃ©bec^H0H 0H0||(819) 999-9999
PV1|1|O|||||199460^TEST2^TEST2^^^^^^^^^^^^^8198888888&Tomate&L|107133^TEST3^TEST3^^^^^^^^^^^^^8195555555&Patate& 
ORC|NW||20231219-169||CM|||||||107133^TEST3^TEST3||||FINAL
OBR|1||20231219-169|8049^MAMMO||20231219141400|20231219142300||||||||SE&Sein|107133^TEST3^TEST3||20231219-169|20231219-16901|||20240103115058||MG|F|||111322^TEST4^TEST4^^^^^^^^^^^^^8192222222&STE-GERTRUDE&L~111322^TEST5^TEST5^^^^^^^^^^^^^8192222222&CLINIQUES Interne&L~101708^TEST6^TEST6^^^^^^^^^^^^^8191111111&CLINIQUE DE SANTE M&L||||1126333&DIEZ MARTINEZ&PATRICIA||BAMA01&BAZIN-CHAMPAGNE&MAUDE|MATJOS01&MATTEAU&JOSÃ‰E-ANNE|20231219141400
```

**Reply by Frazou1** [reply]
``` MSH|^~\&|SRTEXT|Neuron^Radiologie^L|DME|Neuron^Patate^L|20240114161319||ORU^R01|011111|P|2.5.1|||AL||CA|8859/1
PID|1||0003885|AAAA11111111|TEST^TEST|^|19460101|F|||2020, DU TEST^^SAINT^QuÃ©bec^H0H 0H0||(819) 999-9999
PV1|1|O|||||199460^TEST2^TEST2^^^^^^^^^^^^^8198888888&Tomate&L|107133^TEST3^TEST3^^^^^^^^^^^^^8195555555&Patate&L
ORC|NW||20231219-169||CM|||||||107133^TEST3^TEST3||||FINAL
OBR|1||20231219-169|8049^MAMMO||20231219141400|20231219142300||||||||SE&Sein|107133^TEST3^TEST3||20231219-169|20231219-16901|||20240103115058||MG|F|||111322^TEST4^TEST4^^^^^^^^^^^^^8192222222&STE-GERTRUDE&L||||1126333&DIEZ MARTINEZ&PATRICIA||BAMA01&BAZIN-CHAMPAGNE&MAUDE|MATJOS01&MATTEAU&JOSÃ‰E-ANNE|20231219141400


MSH|^~\&|SRTEXT|Neuron^Radiologie^L|DME|Neuron^Patate^L|20240114161319||ORU^R01|011111|P|2.5.1|||AL||CA|8859/1
PID|1||0003885|AAAA11111111|TEST^TEST|^|19460101|F|||2020, DU TEST^^SAINT^QuÃ©bec^H0H 0H0||(819) 999-9999
PV1|1|O|||||199460^TEST2^TEST2^^^^^^^^^^^^^8198888888&Tomate&L|107133^TEST3^TEST3^^^^^^^^^^^^^8195555555&Patate&L
ORC|NW||20231219-169||CM|||||||107133^TEST3^TEST3||||FINAL
OBR|1||20231219-169|8049^MAMMO||20231219141400|20231219142300||||||||SE&Sein|107133^TEST3^TEST3||20231219-169|20231219-16901|||20240103115058||MG|F|||111322^TEST5^TEST5^^^^^^^^^^^^^8192222222&CLINIQUES Interne&L||||1126333&DIEZ MARTINEZ&PATRICIA||BAMA01&BAZIN-CHAMPAGNE&MAUDE|MATJOS01&MATTEAU&JOSÃ‰E-ANNE|20231219141400


MSH|^~\&|SRTEXT|Neuron^Radiologie^L|DME|Neuron^Patate^L|20240114161319||ORU^R01|011111|P|2.5.1|||AL||CA|8859/1
PID|1||0003885|AAAA11111111|TEST^TEST|^|19460101|F|||2020, DU TEST^^SAINT^QuÃ©bec^H0H 0H0||(819) 999-9999
PV1|1|O|||||199460^TEST2^TEST2^^^^^^^^^^^^^8198888888&Tomate&L|107133^TEST3^TEST3^^^^^^^^^^^^^8195555555&Patate&L
ORC|NW||20231219-169||CM|||||||107133^TEST3^TEST3||||FINAL
OBR|1||20231219-169|8049^MAMMO||20231219141400|20231219142300||||||||SE&Sein|107133^TEST3^TEST3||20231219-169|20231219-16901|||20240103115058||MG|F|||101708^TEST6^TEST6^^^^^^^^^^^^^8191&CLINIQUE DE SANTE M&L||||1126333&DIEZ MARTINEZ&PATRICIA||BAMA01&BAZIN-CHAMPAGNE&MAUDE|MATJOS01&MATTEAU&JOSÃ‰E-ANNE|20231219141400
```

**Reply by Frazou1** [reply]
The after should look like this 3 seperates message for 3 obr28 segment

**Reply by pacmano1** [reply]
Make sure that other channel exists on the last line of code.   While you can dispatch multiple message from a source to a destination, I tihnk it is easier to just route to another one.

```JS
for each (var obr28 in msg['OBR']['OBR.28']) {	
	msg['MSH']['MSH.10']['MSH.10.1'] = UUIDGenerator.getUUID()
	msg['OBR']['OBR.28'] = obr28
        router.routeMessage('msg split dumb', SerializerFactory.getSerializer('HL7V2').fromXML(msg));
}
```
This results in 3 messages sent to the downstream channel using your example message you posted here.

**Reply by pacmano1** [reply]
Shortened up previous code a second ago.

**Reply by Frazou1** [reply]
I will try this when i come back to home

**Reply by Frazou1** [reply]
Ok so i can replace all my script  in this 5 lines ? lol  On the last line witch one is the destination one ?

**Reply by pacmano1** [reply]
This channel expects another channel to route to for the new messages.  From that other channel you can route downstream.   The router.routeMessage statement has the name of the channel you need to create.  Obviously, you can create the channel name, whatever you want and then fix the code here.

**Reply by pacmano1** [reply]
And when you post here, do continue to post your code attempts!   It shows you tried and did not just come here asking folks to do your work for you.

I asked you to posta  before and after messages because your code was long I did not want to try to understand it  - I was being lazy.    So seeing the before and after messages is often the fastest way for folks to see what you actually want to happen.

**Reply by pacmano1** [reply]
Also in the code editior in mirth you can do ctrl-space (on a mac) and see code documentation:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/d8c0ce85-2d8f-45f5-8625-0a552df76926)

**Reply by Frazou1** [reply]
Ok i just tried and they work fine !!  WOW. So i did my first channel Call DME_SQL_SOURCE where they receive all the HL7 and i put the script you give me ``` for each (var obr28 in msg['OBR']['OBR.28']) {	
	msg['MSH']['MSH.10']['MSH.10.1'] = UUIDGenerator.getUUID()
	msg['OBR']['OBR.28'] = obr28
        router.routeMessage('DME_SQL', SerializerFactory.getSerializer('HL7V2').fromXML(msg));
} ```   after they resend message to DME_SQL Channel where i have some script to insert data in sql and stuff like that !    Only thing they remain but maybe is already do it.  If a HL7 dhave empty OBR.28 segment or only 1 OBR28 occurence. How the script will act ? Does they will forward the message as well ?

**Reply by pacmano1** [reply]
@Frazou1 , not gonna help you any more if you don't use code formatting.  Fix your last post and your earlier posts.  

Read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by Frazou1** [reply]
Sorry about that, github sucks because i respect the code formatting !! 
![image](https://github.com/nextgenhealthcare/connect/assets/89617565/605a8cea-a314-4d2e-aa8d-a66255dad146)

**Reply by pacmano1** [reply]
Well I feel like you didn't try very hard, lol.  Here is your post formatted and now also quoted:

> Ok i just tried and they work fine !! WOW. So i did my first channel Call DME_SQL_SOURCE where they receive all the HL7 and i put the script you give me:
> ```JS
> for each (var obr28 in msg['OBR']['OBR.28']) {	
> 	msg['MSH']['MSH.10']['MSH.10.1'] = UUIDGenerator.getUUID()
> 	msg['OBR']['OBR.28'] = obr28
>         router.routeMessage('msg split dumb', SerializerFactory.getSerializer('HL7V2').fromXML(msg));
> }
> ```
> after they resend message to DME_SQL Channel where i have some script to insert data in sql and stuff like that ! Only thing they remain but maybe is already do it. If a HL7 dhave empty OBR.28 segment or only 1 OBR28 occurence. How the script will act ? Does they will forward the message as well ?

Well firstly, did you test it with assorted messsages with obr-28s?

I would expect if there is no obr-28 it would forward no messages.  

So just test for that condition and either use the for loop or just router.routeMessage the origninal message.

**Reply by Frazou1** [reply]
Look working !  only thing remain, sometime i have multiple obx segment but i want to keep only one on my hl7
